File: JavaScript/5_DataStructures/7_BinaryHeap.js
class MaxBinayHeap {
  constructor() {
    this.values = [];
  }

  insert(val) {
    /*
    Time Complexity => O(log n)
this function helps to insert new value into heap.
once you insert it to the end of array , it will find its parents and check if its parent value is more
, if its more it breaks or swaps with parent => this happens Recursively.
*/
    this.values.push(val);
    let currentIndex = this.values.length - 1;
    let parentIndex;
    while (currentIndex > 0) {
      parentIndex = Math.floor((currentIndex - 1) / 2);
      if (this.values[parentIndex] >= this.values[currentIndex]) break;
      [this.values[parentIndex], this.values[currentIndex]] = [
        this.values[currentIndex],
        this.values[parentIndex],
      ];
      currentIndex = parentIndex;
    }
    return this;
  }

  extractMax() {
    /*
this Function helps in reteriving the Max value present in the heap in O(1).
here we return value at [0] position since its always MAX and swap the last value to root
and recursively find its childrens and swap if the children are greater than the root.
*/
    let maxValue = this.values[0];
    let parent = this.values.pop();
    let length = this.values.length;
    if (length === 0) return maxValue;

    //Bubble Down
    let parentIndex = 0;
    this.values[parentIndex] = parent;
    let left,
      leftIndex = 1,
      right,
      rightIndex = 2;

    let newIndex = parentIndex;

    while (leftIndex < length || rightIndex < length) {
      left = this.values[leftIndex];
      right = this.values[rightIndex];
      parent = this.values[parentIndex];

      if (left && left > parent) {
        newIndex = leftIndex;
      }
      if (right && right > left && right > parent) {
        newIndex = rightIndex;
      }
      if (newIndex === parentIndex) break;

      [this.values[parentIndex], this.values[newIndex]] = [
        this.values[newIndex],
        this.values[parentIndex],
      ];
      parentIndex = newIndex;
      leftIndex = 2 * parentIndex + 1;
      rightIndex = 2 * parentIndex + 2;
    }
    return maxValue;
  }
}

let BH = new MaxBinayHeap();

/*
Binary Heap is a special trea where only the Maximum value will be as root in
 Max Binary Heap and Minimum in Min Binary Heap.




*/

File: Python/5_DataStructures/7_BinaryHeap.py
import math

class BinaryHeap(object):
	def __init__(self):
		self.heap = []

	def insert(self, val):
		if len(self.heap) == 0:
			return self.heap.append(val)

		self.heap.append(val)

		child_idx = len(self.heap) - 1

		while child_idx > 0:
			parent_idx = math.floor((child_idx-1)/2)
			child = self.heap[child_idx]
			parent = self.heap[parent_idx]

			if parent >= child: break

			self.heap[parent_idx], self.heap[child_idx] = self.heap[child_idx], self.heap[parent_idx]

			child_idx = parent_idx

		return True


	def extractMax(self):
		if len(self.heap) == 0: return False
		if len(self.heap) == 1:
			return self.heap.pop()

		new_root = self.heap.pop()
		max_value = self.heap[0]
		self.heap[0] = new_root
		length = len(self.heap)

		parent_idx = 0
		left_idx, right_idx = 1, 2
		new_idx = 0

		while True:
			parent = self.heap[parent_idx]

			if left_idx >= length: break
			left = self.heap[left_idx]
			if(left > parent):
				new_idx = left_idx

			if right_idx >= length: break
			right = self.heap[right_idx]
			if(right > left and right > parent):
				new_idx = right_idx

			if new_idx == parent_idx: break

			self.heap[new_idx], self.heap[parent_idx] = self.heap[parent_idx], self.heap[new_idx]

			parent_idx = new_idx
			left_idx = (2 * parent_idx) + 1
			right_idx = (2 * parent_idx) + 2

		return max_value